File: graduacao/orientacao_objetos/trabalho-ii/Artjr.cpp
#include "Artjr.hpp"

namespace producoes {

    Artjr::Artjr(const string& titulo, const string& idioma, const string& cidade,
      const int& paginas, const int& tipo, const time_t& data_, const string& issn_)
      : Producao(titulo, idioma, cidade, "", paginas, tipo),
      issn(issn_), data(data_) {}

    void Artjr::print(ostream& out) const {
        out << titulo << ";" << idioma << ";" << cidade
            << ";" << formatDate(data, DATE_FORMAT_PT_BR_SHORT)
            << ";" << issn << ";";
        //Verifica se paginas sera impresso
        if(paginas > 0) out << paginas;
    }

    bool Artjr::operator<(const Producao& ref) {
        //Converte para Artjr para poder acessar os dados
        Artjr& o = (Artjr&) ref;
        //Compara os dados individualmente
        int cmp = titulo.compare(o.titulo);
        if(cmp<0) return true; else if(cmp>0) return false;
        cmp = idioma.compare(o.idioma);
        if(cmp<0) return true; else if(cmp>0) return false;
        cmp = cidade.compare(o.cidade);
        if(cmp<0) return true; else if(cmp>0) return false;
        cmp = data - o.data;
        if(cmp<0) return true; else if(cmp>0) return false;
        cmp = issn.compare(o.issn);
        if(cmp<0) return true; else if(cmp>0) return false;
        cmp = paginas - o.paginas;
        if(cmp<0) return true; else return false;
    }

}

File: graduacao/orientacao_objetos/trabalho-i/Producoes/Artjr.java
package Producoes;
import java.text.DateFormat;
import java.text.SimpleDateFormat;
import java.util.Date;

/**
 * Producao que representa producoes do tipo Artigo em
 * jornais e revistas.
 * 
 * @author Gabriel S. Baptista
 *
 */
public class Artjr extends Producao {
	private Date dataDePublicacao;
	private String issn;

	/**
	 * Construtor da Classe
	 * 
	 * @param titulo String representando o titulo.
	 * @param idioma String representando o idioma.
	 * @param cidade String representando a cidade.
	 * @param paginas Inteiro representando a quantidade de paginas.
	 * @param data Date representando a data da producao.
	 * @param issn String representando o issn do documento.
	 */
	public Artjr(String titulo, String idioma, String cidade, int paginas, Date data, String issn) {
		super(titulo, idioma, cidade, "", paginas);
		this.dataDePublicacao = data;
		this.issn = issn;
	}

	/**
	 * Imprime o cabecalho da Classe artjr.
	 */
	public static void imprimirCabecalho() {
		System.out.println("Titulo;Idioma;Cidade;Data;ISSN;Paginas");
	}
	
	/**
	 * Sobreescreve o metodo toString, ou seja, representa
	 * a maneira usual de se exibir uma producao do tipo artjr.
	 */
	@Override
	public String toString() {
		String texto = this.getTitulo() + ";" + this.getIdioma() + ";" + this.getCidade() + ";" + this.getData() + ";"
				+ this.issn + ";";
		if (this.getPaginas() > 0)
			texto += this.getPaginas();
		return texto;
	}

	/**
	 * Retorna a data do artigo em jornais e revistas atual.
	 * 
	 * @return data do documento atual em string formatado na saida brasileira.
	 */
	private String getData() {
		DateFormat formater = new SimpleDateFormat("dd/MM/yyyy");
		return formater.format(this.dataDePublicacao);
	}

	/**
	 * Sobreescreve o metodo equals.
	 */
	@Override
	public boolean equals(Object obj) {
		if (!(obj instanceof Artjr))
			return false;
		Artjr referencia = (Artjr) obj;
		if (this.compareTo(referencia) == 0) {
			return true;
		} else {
			return false;
		}
	}

	/**
	 * Sobreescreve o metodo compareTo, ou seja,  representa
	 * a maneira usual de se comparar producoes do tipo artrj.
	 */
	@Override
	public int compareTo(Producao o) {
		Artjr referencia = (Artjr) o;
		// Verifica as caracteristicas
		int cmp = this.getTitulo().compareTo(referencia.getTitulo());
		if (cmp != 0) return cmp;
		cmp = this.getIdioma().compareTo(referencia.getIdioma());
		if (cmp != 0) return cmp;
		cmp = this.getCidade().compareTo(referencia.getCidade());
		if (cmp != 0) return cmp;
		cmp = this.dataDePublicacao.compareTo(referencia.dataDePublicacao);
		if (cmp != 0) return cmp;
		cmp = this.issn.compareTo(referencia.issn);
		if (cmp != 0) return cmp;
		return this.getPaginas() - referencia.getPaginas();
	}

}
